add a custom alternative strategy

MichaelMure opened the pull request.

MichaelMure committed:
add a custom alternative strategy

ncocchiaro commented on strategy_infura.go:
```diff
@@ -0,0 +1,443 @@
+
+	// protect the new instances
+	fmt.Printf("Enable instance protection on the new instances\n")
+	err = setInstanceProtection(ctx, sess, asgName, newInstances)
```
You shouldn't have to do this if the ASG is set with scale in protection by default. By this point if the instances are not already protected, the ASG may have already selected them for replacement.

MichaelMure replied:
I wanted to avoid making assumption and also support ASG without scale in protection by default, as for go-ipfs nodes. Also this is a public repo so let's not even assume Infura ;)

One thing I can do is to start the new instances with scale in protection.

MichaelMure replied:
Also, I should probably detect how the ASG is configured regarding scale in protection to put back the things how they were before.

ncocchiaro commented on strategy_infura.go:
```diff
@@ -0,0 +1,443 @@
+	// get back to normal size, cull the old instances
+	fmt.Printf("Set back the ASG size:\n\tmaxSize %d-->%d\n\tdesiredCapacity %d-->%d\n",
+		2*originalAsg.maxSize, originalAsg.maxSize, 2*originalAsg.desiredCapacity, originalAsg.desiredCapacity)
+	err = setASGCapacities(ctx, sess, asgName, originalAsg.maxSize, originalAsg.desiredCapacity)
```
This would be best to do before the scale in protection removal step above, or the ASG may start an AZ rebalancing process and start different instances to replace those that are no longer protected (rebalancing is paused permanently on most hammerhead ASGs, but not guaranteed to be that way for others).

MichaelMure replied:
I believe this is not a problem anymore with the new structure. Instances can be terminated,replaced ... during the refresh. Healthy and ready instances will be protected until the process end.

ncocchiaro commented on strategy_infura.go:
```diff
@@ -0,0 +1,443 @@
+
+	// unprotect the new instances
+	fmt.Printf("Remove instance protection on the new instances\n")
+	err = removeInstanceProtection(ctx, sess, asgName, newInstances)
```
Any reason we want to not just leave the scale in protection in place permanently?

MichaelMure replied:
To not assume anything about the ASG. There might be autoscaling rules there so leaving the protection might mess with that.

ncocchiaro replied:
Fair to support the general case.

ryanschneider commented on strategy_aws.go:
```diff
@@ -0,0 +1,83 @@
+package main
```
very minor:  I learned from Andrey that `_` is special in Go file names (which is why tests have to be in `_test.go`.  Basically `_foo` means it'll be included if the `foo` build flag is set.  ever since then I've tried to avoid underscores in filenames.  https://stackoverflow.com/a/39625565

MichaelMure replied:
oof that's sneaky.

MichaelMure replied:
Changed for dashes

ncocchiaro commented on strategy_infura.go:
```diff
@@ -0,0 +1,443 @@
+				continue
+			}
+
+			// if we don't have a target group, end the checks here
```
This may be another area to revisit in the future, as we'd still need a target group-style health check to know if the instances are effectively healthy (the ASG health check is separate). For example, if we used this on an emergency hammerhead ASG not tied to any TG, we'd take out the old instances before the new ones are healthy from an eth1 service point of view.

MichaelMure replied:
so that would be an option to pass an arbitrary health check ? There would be no way to detect that, right?

ncocchiaro replied:
Yeah I think in the future along with custom strategies we may want the ability to compose in a custom health check as well.

ryanschneider commented on strategy_infura.go:
```diff
@@ -0,0 +1,443 @@
+		return err
+	}
+
+	fmt.Printf("Refresh (strategy: infura-refresh) started on ASG %s\n", asgName)
```
minor: using `log` instead of `fmt` gives you timestamps which are probably useful for long-running strategies like this.

ncocchiaro replied:
Would be really good to have timestamps for all messages or all steps.

ncocchiaro reviewed (commented):
This is great, I took the liberty to leave some high-level comments. Main thing is to use the TG health check when possible, which this already does, but there may be some other undesired cases this can run into.

MichaelMure committed:
use `log` to have proper timestamping

MichaelMure committed:
respect original scale-in protection

MichaelMure commented:
Successful run on a scale-in protected ASG:
```
2020/12/07 18:38:18 Refresh started (asg: ipfs-v6-dev-od, region: us-east-1, strategy: infura-refresh)
2020/12/07 18:38:18 ASG protect instances by default: true
2020/12/07 18:38:18 Existing instances:
2020/12/07 18:38:18     i-018682b1d445352b6 lifecycle:InService asg-based-health:Healthy
2020/12/07 18:38:18     i-04e7c6adca76b93e6 lifecycle:InService asg-based-health:Healthy
2020/12/07 18:38:18     i-066b3d7e8462cba77 lifecycle:InService asg-based-health:Healthy
2020/12/07 18:38:18     i-07003ddd394600aa0 lifecycle:InService asg-based-health:Healthy
2020/12/07 18:38:18     i-07cf361d116399f58 lifecycle:InService asg-based-health:Healthy
2020/12/07 18:38:18     i-09742fcbaaedc4463 lifecycle:InService asg-based-health:Healthy
2020/12/07 18:38:18     i-0b16877fbd3090d5a lifecycle:InService asg-based-health:Healthy
2020/12/07 18:38:18     i-0b5d0ff407c5d8aa8 lifecycle:InService asg-based-health:Healthy
2020/12/07 18:38:18     i-0eda59d82726b41cc lifecycle:InService asg-based-health:Healthy
2020/12/07 18:38:18     i-0f3b0d8e069ab98c5 lifecycle:InService asg-based-health:Healthy
2020/12/07 18:38:18 Enable instance protection on old instances
2020/12/07 18:38:18 Double the ASG size:
2020/12/07 18:38:18     maxSize 10-->20
2020/12/07 18:38:18     desiredCapacity 10-->20
2020/12/07 18:38:18 Detect new instances and wait for them to be healthy
2020/12/07 18:38:34     found new instance: i-011df7fd01bfab0d6 lifecycle:Pending asg-based-health:Healthy
2020/12/07 18:38:34     found new instance: i-02e74e5866328a93b lifecycle:Pending asg-based-health:Healthy
2020/12/07 18:38:34     found new instance: i-040031477857130dc lifecycle:Pending asg-based-health:Healthy
2020/12/07 18:38:35     found new instance: i-0606def8bb461d053 lifecycle:Pending asg-based-health:Healthy
2020/12/07 18:38:35     found new instance: i-07824873931aa96ef lifecycle:Pending asg-based-health:Healthy
2020/12/07 18:38:35     found new instance: i-0875d180dc1824321 lifecycle:Pending asg-based-health:Healthy
2020/12/07 18:38:35     found new instance: i-08f93c0a1f938a893 lifecycle:Pending asg-based-health:Healthy
2020/12/07 18:38:35     found new instance: i-0a9273dcf7c5d4f1c lifecycle:Pending asg-based-health:Healthy
2020/12/07 18:38:35     found new instance: i-0c9367bae49562c12 lifecycle:Pending asg-based-health:Healthy
2020/12/07 18:38:36     found new instance: i-0db0536f7b5c01c6c lifecycle:Pending asg-based-health:Healthy
2020/12/07 18:39:07     (1/10) Instance i-011df7fd01bfab0d6 is ready
2020/12/07 18:39:07     (2/10) Instance i-02e74e5866328a93b is ready
2020/12/07 18:39:07     (3/10) Instance i-040031477857130dc is ready
2020/12/07 18:39:07     (4/10) Instance i-0606def8bb461d053 is ready
2020/12/07 18:39:07     (5/10) Instance i-07824873931aa96ef is ready
2020/12/07 18:39:07     (6/10) Instance i-0875d180dc1824321 is ready
2020/12/07 18:39:07     (7/10) Instance i-08f93c0a1f938a893 is ready
2020/12/07 18:39:07     (8/10) Instance i-0a9273dcf7c5d4f1c is ready
2020/12/07 18:39:07     (9/10) Instance i-0c9367bae49562c12 is ready
2020/12/07 18:39:07     (10/10) Instance i-0db0536f7b5c01c6c is ready
2020/12/07 18:39:07 Remove instance protection on the old instances
2020/12/07 18:39:07 Set back the ASG size:
2020/12/07 18:39:07     maxSize 20-->10
2020/12/07 18:39:07     desiredCapacity 20-->10
2020/12/07 18:39:08 Wait for scaling in
2020/12/07 18:39:13     Instance count: 20
2020/12/07 18:39:18     Instance count: 20
2020/12/07 18:39:23     Instance count: 20
2020/12/07 18:39:28     Instance count: 20
2020/12/07 18:39:34     Instance count: 20
2020/12/07 18:39:39     Instance count: 20
2020/12/07 18:39:44     Instance count: 20
2020/12/07 18:39:49     Instance count: 20
2020/12/07 18:39:55     Instance count: 16
2020/12/07 18:40:00     Instance count: 16
2020/12/07 18:40:05     Instance count: 16
2020/12/07 18:40:10     Instance count: 16
2020/12/07 18:40:15     Instance count: 11
2020/12/07 18:40:21     Instance count: 11
2020/12/07 18:40:26     Instance count: 11
2020/12/07 18:40:31     Instance count: 11
2020/12/07 18:40:37     Instance count: 10
2020/12/07 18:40:37 Done in 2m19.485929632s.
```

Run with a faked error when waiting for the scaling in:
```
2020/12/07 18:54:48 Refresh started (asg: ipfs-v6-dev-od, region: us-east-1, strategy: infura-refresh)
2020/12/07 18:54:48 ASG protect instances by default: true
2020/12/07 18:54:48 Existing instances:
2020/12/07 18:54:48     i-0052995c8baec4227 lifecycle:InService asg-based-health:Healthy
2020/12/07 18:54:48     i-059379dca35ff5399 lifecycle:InService asg-based-health:Healthy
2020/12/07 18:54:48     i-05da36ebd96f7c485 lifecycle:InService asg-based-health:Healthy
2020/12/07 18:54:48     i-066e2bc0bf8b4dfc7 lifecycle:InService asg-based-health:Healthy
2020/12/07 18:54:48     i-0880d8e26276c8902 lifecycle:InService asg-based-health:Healthy
2020/12/07 18:54:48     i-0ad33c81be5642b90 lifecycle:InService asg-based-health:Healthy
2020/12/07 18:54:48     i-0af24005b7370c4c2 lifecycle:InService asg-based-health:Healthy
2020/12/07 18:54:48     i-0cd6c20836cb42c90 lifecycle:InService asg-based-health:Healthy
2020/12/07 18:54:48     i-0d7b2a3439d16d3c0 lifecycle:InService asg-based-health:Healthy
2020/12/07 18:54:48     i-0ef503a929fd87ff2 lifecycle:InService asg-based-health:Healthy
2020/12/07 18:54:48 Enable instance protection on old instances
2020/12/07 18:54:48 Double the ASG size:
2020/12/07 18:54:48     maxSize 10-->20
2020/12/07 18:54:48     desiredCapacity 10-->20
2020/12/07 18:54:48 Detect new instances and wait for them to be healthy
2020/12/07 18:55:09     found new instance: i-01daf6de65b465c4b lifecycle:Pending asg-based-health:Healthy
2020/12/07 18:55:09     found new instance: i-02186a97ac77659cd lifecycle:Pending asg-based-health:Healthy
2020/12/07 18:55:10     found new instance: i-04476710e0fd2fa13 lifecycle:Pending asg-based-health:Healthy
2020/12/07 18:55:10     found new instance: i-048487694d93a8419 lifecycle:Pending asg-based-health:Healthy
2020/12/07 18:55:10     found new instance: i-0828c9cf0e5b39477 lifecycle:Pending asg-based-health:Healthy
2020/12/07 18:55:10     found new instance: i-0ace3fcd1d8d87906 lifecycle:Pending asg-based-health:Healthy
2020/12/07 18:55:11     found new instance: i-0ce3726a3b0423581 lifecycle:Pending asg-based-health:Healthy
2020/12/07 18:55:11     found new instance: i-0e8b92ac94dbf317d lifecycle:Pending asg-based-health:Healthy
2020/12/07 18:55:11     found new instance: i-0f4f16ec5f3abd8dc lifecycle:Pending asg-based-health:Healthy
2020/12/07 18:55:11     found new instance: i-0fa2d676de27072ae lifecycle:Pending asg-based-health:Healthy
2020/12/07 18:55:42     (1/10) Instance i-01daf6de65b465c4b is ready
2020/12/07 18:55:42     (2/10) Instance i-02186a97ac77659cd is ready
2020/12/07 18:55:42     (3/10) Instance i-04476710e0fd2fa13 is ready
2020/12/07 18:55:42     (4/10) Instance i-048487694d93a8419 is ready
2020/12/07 18:55:42     (5/10) Instance i-0828c9cf0e5b39477 is ready
2020/12/07 18:55:42     (6/10) Instance i-0ace3fcd1d8d87906 is ready
2020/12/07 18:55:42     (7/10) Instance i-0ce3726a3b0423581 is ready
2020/12/07 18:55:42     (8/10) Instance i-0e8b92ac94dbf317d is ready
2020/12/07 18:55:42     (9/10) Instance i-0f4f16ec5f3abd8dc is ready
2020/12/07 18:55:42     (10/10) Instance i-0fa2d676de27072ae is ready
2020/12/07 18:55:42 Remove instance protection on the old instances
2020/12/07 18:55:43 Set back the ASG size:
2020/12/07 18:55:43     maxSize 20-->10
2020/12/07 18:55:43     desiredCapacity 20-->10
2020/12/07 18:55:43 Wait for scaling in
2020/12/07 18:55:43 ======================================================
2020/12/07 18:55:43 ROLLBACK due to error: fake error
2020/12/07 18:55:43 ======================================================
2020/12/07 18:55:43 Wait 10 min for the ASG to stabilize
fake error
```

MichaelMure committed:
implement rollback

MichaelMure committed:
handle instance failure at launch by locking new healthy instances later

MichaelMure commented:
Done since the last review:
- proper timestamps in the logs
- respect the original scale-in protection of the ASG
- implement rollback
- handle instance failure at launch by continuously detecting new/healthy instances, protecting them only when healthy, lock and scaling down when the expected count is there. This leave the opportunity for the ASG to replace failed instances.
- ironing out and bug fixes

MichaelMure commented:
Possible work left:
- adjustable or auto-tuned strategy timeout
- optional custom health-check for deployment without a TG to avoid swapping the instances before them actually being ready

MichaelMure reviewed (commented):
Please github, I don't want to review my own PR right now.

ncocchiaro commented on strategy_infura.go:
```diff
@@ -0,0 +1,524 @@
+			return nil, ctx.Err()
+		case err := <-chanErr2:
+			if err != nil {
+				log.Printf("\terror while polling autoscaling or target group reported target health: %v", err)
```
In this case this instance may be caught at the next polling iteration, correct?

MichaelMure replied:
That's correct. This polling loop only end when the context end. The error is only logged, which means that the next loop can find zero, one or more new instances.

MichaelMure replied:
To be more clear as it's not so obvious on a github PR: there is **four** endless loop binded the the context lifecycle here:
- one in `detectNewInstances()` doing ASG polling, which output to two **channels** for 1) values and 2) errors
- the goroutine in this function that log and discards the errors from the error channel (line 294)
- the value channel get passed to `detectInstancesReady()` which poll the ASG health check and possibly the TG. This function also output to two channels for values and errors
- this `for` loop that 1) log and discard errors from the second error channel and 2) read from the value channel, set the instance protection and accumulate the healthy instances until we have the expected count

ncocchiaro reviewed (approved):
It's a large PR and with these flows it's tough to catch all edge cases a priori, but this should be a good start to iterate on as needed.

MichaelMure committed:
add comments and documentation

ryanschneider commented on asg-instance-refresh.go:
```diff
@@ -1,98 +1,63 @@
+var ErrStrategyFailed = errors.New("strategy failed")
 
 func main() {
+	region := flag.String("region", "us-east-1", "The AWS region. Default to us-east-1.")
```
Minor/comment: if like me you get sick of typing `*region` everywhere you can use `var region string; flag.StringVar(&region, ....)`.

Not suggesting you make this change as part of this PR just a TIL.

MichaelMure replied:
if I get that sick of it, I'll move to `cobra` or something like that ;)

ryanschneider commented on strategy_infura_refresh.go:
```diff
@@ -0,0 +1,547 @@
+	// - new instances have the protection enabled
+	// For the rollback:
+	// - wait 10 minutes to give the ASG a chance to stabilize with keeping the new instances and removing the old ones
+	// - make sure to remove have the protection set as desired on the new instances
```
```suggestion
	// - make sure to have the protection set as desired on the new instances
```

MichaelMure committed:
don't use underscores in file names

https://stackoverflow.com/questions/25161774/what-are-conventions-for-filenames-in-go/39625565#39625565

ryanschneider commented on strategy_infura_refresh.go:
```diff
@@ -0,0 +1,547 @@
+			select {
+			case <-ctx.Done():
+				return
+			case inst := <-newInstances:
```
Once `newInstances` is closed by `detectNewInstances` won't this continuously return a zero-value instance object?

For example: https://play.golang.org/p/9ZLMeNYlNkY

MichaelMure replied:
Yeah good point. I suppose most of the time `<-ctx.Done()` would trigger first and exist but that's a race at least.

ryanschneider reviewed (commented):
Overall lgtm but I think there's some potential for unexpected behavior in `detectInstancesReady` once the `newInstances` channel is closed.

I _think_ that since we never `range` on the channels the simplest fix would be to just never close them (you don't _need_ to close channels they get GC'd just fine once no one points to them even if still open).

MichaelMure committed:
fix a race when closing channels

ryanschneider reviewed (approved):
fix lgtm!
